src/quic/application.cc

```diff
@@ -0,0 +1,447 @@
+                                stream_data.count,
+                                uv_hrtime());
+  return ret;
+}
+
+// The DefaultApplication is the default implementation of Session::Application
+// that is used for all unrecognized ALPN identifiers.
+class DefaultApplication final : public Session::Application {
+ public:
+  // Marked NOLINT because the cpp linter gets confused about this using
+  // statement not being sorted with the using v8 statements at the top
+  // of the namespace.
+  using Application::Application;  // NOLINT
+
+  bool ReceiveStreamData(Stream* stream,
+                         const uint8_t* data,
+                         size_t datalen,
+                         Stream::ReceiveDataFlags flags) override {
+    DCHECK_NOT_NULL(stream);
+    if (!stream->is_destroyed()) stream->ReceiveData(data, datalen, flags);
+    return true;
+  }
+
+  int GetStreamData(StreamData* stream_data) override {
+    DCHECK_NOT_NULL(stream_data);
+    // If the queue is empty, there aren't any streams with data yet
+    if (stream_queue_.IsEmpty()) return 0;
+
+    const auto get_length = [](auto vec, size_t count) {
+      CHECK_NOT_NULL(vec);
+      size_t len = 0;
+      for (size_t n = 0; n < count; n++) len += vec[n].len;
+      return len;
+    };
+
+    Stream* stream = stream_queue_.PopFront();
+    CHECK_NOT_NULL(stream);
+    stream_data->stream.reset(stream);
+    stream_data->id = stream->id();
+    auto next =
+        [&](int status, const ngtcp2_vec* data, size_t count, bob::Done done) {
+          switch (status) {
+            case bob::Status::STATUS_BLOCK:
+              // Fall through
+            case bob::Status::STATUS_WAIT:
+              return;
+            case bob::Status::STATUS_EOS:
+              stream_data->fin = 1;
+          }
+
+          stream_data->count = count;
+
+          if (count > 0) {
+            stream->Schedule(&stream_queue_);
+            stream_data->remaining = get_length(data, count);
+          } else {
+            stream_data->remaining = 0;
+          }
+
+          // Not calling done here because we defer committing
+          // the data until after we're sure it's written.
+        };
+
+    if (LIKELY(!stream->is_eos())) {
+      int ret = stream->Pull(std::move(next),
+                             bob::Options::OPTIONS_SYNC,
+                             stream_data->data,
+                             arraysize(stream_data->data),
+                             kMaxVectorCount);
+      switch (ret) {
+        case bob::Status::STATUS_EOS:
+          stream_data->fin = 1;
+          break;
+      }
+    } else {
+      stream_data->fin = 1;
+    }
+
+    return 0;
+  }
+
+  void ResumeStream(int64_t id) override { ScheduleStream(id); }
+
+  bool ShouldSetFin(const StreamData& stream_data) override {
+    auto const is_empty = [](auto vec, size_t cnt) {
+      size_t i;
+      for (i = 0; i < cnt && vec[i].len == 0; ++i) {
+      }
```

Comment:
Is this for loop intended to be empty?

Reply:
I think so. Something like `all_of(vec, [](&v) -> { return v.len == 0; })`.

Reply:
What about `std::find`?

`std::find(vec.begin(), vec.end(), [](auto it) { return it.len != 0 })`

Reply:
Yes it's intentional. From what I've seen the perf characteristics of this vs. all_of or std::find are generally the same.